### PR TITLE
[graph] Changed graph info json output

### LINKED ISSUE #13862: [question] Conan2: Reading / dumping version ranges for dependencies ?
### What is your question?

In order to integrate a Coan2 into our existing CI infrastructure, we would like to print (or dump to file) not only the RESOLVED dependencies  (such as those found in self.dependencies) but also the requirement as set in the requirements() method, that could be a version range, such as:
`self.test_requires("gtest/[>1]")`

It seems this information is available in the dependency graph (and is printed by Conan2 under "Resolved version ranges") but the CLI command "`conan graph info`" does not display it.

Is it mandatory to use the graph API ? I'd rather not use it as it's marked as unstable for now.

I've searched through the existing issues and I haven't found what I'm looking for.

Thanks for any help.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

### LINKED ISSUE #13911: Discuss graph info json output
Things to do:

- Convert ``nodes`` list to dict {id: nodecontents}
- Add the metadata author, license, homepage, etc always, even if ``null``
- Convert the ``requires`` to a full serialization of ``transitive_deps`` in the form: ``{id: {ref: rref, build=True, libs=False, ...``
- Add summary of graph resolution, similar to command terminal stdout for ``resolved-ranges``, ``overrides``, etc.

### PR DESCRIPTION
**Pull request by franramirez688**
Changelog: Bugfix: Fixed and completed the `graph info xxxx --format json` output, to publicly document it.
Docs: https://github.com/conan-io/docs/pull/3236

Closes: https://github.com/conan-io/conan/issues/13911
Closes: https://github.com/conan-io/conan/issues/13862

For instance:

```
{
    "nodes": {
         "0": {
            "ref": "pkga/0.1",
            "id": 0,
            "recipe": "Consumer",
            "package_id": "041db7e16de982fbd01f61f3809fc1c206fa9931",
            "prev": null,
            "build_id": null,
            "binary": null,
            "invalid_build": false,
            "info_invalid": null,
            "name": "pkga",
            "user": null,
            "channel": null,
            "url": null,
            "license": null,
            "author": null,
            "description": null,
            "homepage": null,
            "build_policy": null,
            "upload_policy": null,
            "revision_mode": "hash",
            "provides": null,
            "deprecated": null,
            "win_bash": null,
            "win_bash_run": null,
            "default_options": null,
            "options_description": null,
            "version": "0.1",
            "topics": null,
            "package_type": "unknown",
            "settings": {},
            "options": {},
            "options_definitions": {},
            "generators": [],
            "system_requires": {},
            "recipe_folder": "/private/var/folders/s8/zgn806y90xq7bdh64hhfw9th0000gr/T/tmpze4f_1t9conans/path with spaces",
            "source_folder": null,
            "build_folder": null,
            "generators_folder": null,
            "package_folder": null,
            "cpp_info": {
                "root": {
                    "includedirs": [
                        "include"
                    ],
                    "srcdirs": null,
                    "libdirs": [
                        "lib"
                    ],
                    "resdirs": null,
                    "bindirs": [
                        "bin"
                    ],
                    "builddirs": null,
                    "frameworkdirs": null,
                    "system_libs": null,
                    "frameworks": null,
                    "libs": null,
                    "defines": null,
                    "cflags": null,
                    "cxxflags": null,
                    "sharedlinkflags": null,
                    "exelinkflags": null,
                    "objects": null,
                    "sysroot": null,
                    "requires": null,
                    "properties": null
                }
            },
            "label": "conanfile.py (pkga/0.1)",
            "dependencies": {
                "1": {
                    "ref": "pkgb/0.1@common/unstable",
                    "run": "False",
                    "libs": "True",
                    "skip": "False",
                    "test": "False",
                    "force": "False",
                    "direct": "True",
                    "build": "False",
                    "transitive_headers": "None",
                    "transitive_libs": "None",
                    "headers": "True",
                    "package_id_mode": "semver_mode",
                    "visible": "True"
                },
                "3": {
                    "ref": "ros_perception/1.1.4@3rdparty/unstable",
                    "run": "False",
                    "libs": "True",
                    "skip": "False",
                    "test": "False",
                    "force": "False",
                    "direct": "True",
                    "build": "False",
                    "transitive_headers": "None",
                    "transitive_libs": "None",
                    "headers": "True",
                    "package_id_mode": "semver_mode",
                    "visible": "True"
                },
                "2": {
                    "ref": "ros_core/pr-53@3rdparty/snapshot",
                    "run": "False",
                    "libs": "True",
                    "skip": "False",
                    "test": "False",
                    "force": "True",
                    "direct": "True",
                    "build": "False",
                    "transitive_headers": "None",
                    "transitive_libs": "None",
                    "headers": "True",
                    "package_id_mode": "semver_mode",
                    "visible": "True"
                }
            },
            "context": "host",
            "test": false
    },

    [....],

     "root": {
        "0": "pkga/0.1"
    },
    "overrides": {
        "ros_core/[~1.1]@3rdparty/unstable": [
            "ros_core/pr-53@3rdparty/snapshot"
        ],
        "ros_core/1.1.4@3rdparty/unstable": [
            "ros_core/pr-53@3rdparty/snapshot#4d670581ccb765839f2239cc8dff8fbd"
        ]
    },
    "resolved-ranges": {
        "pkgb/[~0]@common/unstable": "pkgb/0.1@common/unstable",
        "ros_perception/[~1.1]@3rdparty/unstable": "ros_perception/1.1.4@3rdparty/unstable"
    }
}
```

### CODE REVIEW

File: conans/client/graph/graph.py [hunk 1]
```diff
@@ -208,6 +208,15 @@ def __repr__(self):
         return repr(self.conanfile)
 
     def serialize(self):
+        def _topics():
```
Comment (AbrilRBS) [inline conans/client/graph/graph.py]: I remember changing ConanFile so topics would always be iterable. Did you get bitten by this?

File: conans/client/graph/graph.py [hunk 2]
```diff
@@ -224,7 +238,8 @@ def serialize(self):
         result.update(self.conanfile.serialize())
         result["context"] = self.context
         result["test"] = self.test
-        result["requires"] = {n.id: n.ref.repr_notime() for n in self.neighbors()}
+        result["transitive_deps"] = {d.node.id: d.require.serialize()
```
Comment (memsharded) [inline conans/client/graph/graph.py]: Maybe we want to call it ``requires``? ``transitive_requires``? ``dependencies``?

Comment (franramirez688) [reply conans/client/graph/graph.py]: Yeah, I wanted to bring it here to discuss the best name. Perhaps, keeping `requires` is good enough? wdyt @czoido @RubenRBS ?

File: conans/client/graph/graph.py (hunk 1 above)
Comment (franramirez688) [reply conans/client/graph/graph.py]: Good point! I just reviewed the code, and you're completely right! Thanks!

File: conans/client/graph/graph.py (hunk 2 above)
Comment (AbrilRBS) [reply conans/client/graph/graph.py]: I'd say `dependencies`, for symmetry with `self.dependencies` in a ConanFile

Comment (memsharded) [reply conans/client/graph/graph.py]: Yes, ``dependencies`` could be the best, it is the closest, it is how ``dependencies`` is computed:
```python
 def from_node(node):
        d = OrderedDict((require, ConanFileInterface(transitive.node.conanfile))
                        for require, transitive in node.transitive_deps.items())
        return ConanFileDependencies(d)
```
It is basically a thin adaptation over ``transitive_deps``, with an interface to protect internal data on top of it (we might want to make accessible some of this data over the ``ConanFileInterface`` if necessary and requested by users)

**Review by AbrilRBS — APPROVED** [review]

File: conans/test/integration/command/create_test.py [hunk 3]
```diff
@@ -370,7 +369,7 @@ def test_create_format_json():
                  ....
                 },
                 {'ref': 'hello/0.1#18d5440ae45afc4c36139a160ac071c7',
-                 'requires': {'2': 'pkg/0.2#db78b8d06a78af5c3ac56706f133098d'},
+                 'dependencies': {'2': 'pkg/0.2#db78b8d06a78af5c3ac56706f133098d'},
```
Comment (memsharded) [inline conans/test/integration/command/create_test.py]: This doesn't look good, the inner thing should be a dict, not a reference, shouldn't it?

**Review by memsharded — APPROVED** [review]
Looks good, just a minor comment

File: conans/test/integration/command/create_test.py (hunk 3 above)
Comment (franramirez688) [reply conans/test/integration/command/create_test.py]: This is a docstring. That's the reason why the test is passing. Let me update it.

**Review by czoido — APPROVED** [review]

File: conan/cli/formatters/graph/graph_info_text.py [hunk 4]
```diff
@@ -31,7 +31,7 @@ def format_graph_info(result):
     out.title("Basic graph information")
     serial = graph.serialize()
     serial = filter_graph(serial, package_filter, field_filter)
-    for n in serial["nodes"]:
+    for _, n in serial["nodes"].items():
```
Comment (AbrilRBS) [inline conan/cli/formatters/graph/graph_info_text.py]:
```suggestion
    for n in serial["nodes"].values():
```

**Review by AbrilRBS — APPROVED** [review]

**Pull request merged**